### PR TITLE
Leaf localization

### PR DESCRIPTION
**Pull request by NathanFlurry**
Adds `localize` tag to all Leaf documents that utilizes the Droplet provided localization. You can use this like `#localize("en", "welcome", "title")`.

### CODE REVIEW

File: Sources/Vapor/Localization/Localization.swift [hunk 1]
```diff
@@ -1,26 +1,52 @@
...
 public class Localization {
-    private let localization: JSONDirectory
+    private let localizations: [String: Node]
+    private let defaultLocale: String
 
-    public convenience init(workingDirectory: String) throws {
-        let configDirectory = workingDirectory.finished(with: "/") + "Localization/"
-        let localization = try FileManager.loadDirectory(configDirectory)
-        self.init(jsonDirectory: localization)
+    public convenience init(workingDirectory: String, path: String? = nil) throws {
```
Comment (LoganWright) [inline Sources/Vapor/Localization/Localization.swift]: My vote on this one is to leave the the path out of it and just have `localizationDirectory`.  Where Droplet adds `/Localization` as a default. I've found it a little more straight forward and people have been less surprised by the result.

File: Sources/Vapor/Localization/Localization.swift [hunk 2]
```diff
@@ -1,26 +1,52 @@
...
+        let localizationDirectory = workingDirectory.finished(with: "/") +  (path ?? "Localization/")
+        try self.init(path: localizationDirectory)
+    }
+    
+    public convenience init(path: String) throws {
+        // List the files in the directory or have no localizations if could not find files
+        guard let contents = try? FileManager.contentsOfDirectory(path) else {
```
Comment (LoganWright) [inline Sources/Vapor/Localization/Localization.swift]: outer method throws, I think we can use just `try` here and propagate upwards

File: Sources/Vapor/Localization/Localization.swift [hunk 3]
```diff
@@ -1,26 +1,52 @@
...
-    public convenience init(workingDirectory: String) throws {
-        let configDirectory = workingDirectory.finished(with: "/") + "Localization/"
-        let localization = try FileManager.loadDirectory(configDirectory)
-        self.init(jsonDirectory: localization)
+    public convenience init(workingDirectory: String, path: String? = nil) throws {
+        let localizationDirectory = workingDirectory.finished(with: "/") +  (path ?? "Localization/")
+        try self.init(path: localizationDirectory)
+    }
+    
+    public convenience init(path: String) throws {
```
Comment (LoganWright) [inline Sources/Vapor/Localization/Localization.swift]: `path` => `localizationDirectory`. (probably omitted when above initializer changes.)

File: Sources/Vapor/Localization/Localization.swift [hunk 4]
```diff
@@ -1,26 +1,52 @@
...
+        }
+        
+        self.init(localizations: localizations)
     }
 
-    init(jsonDirectory: JSONDirectory? = nil) {
-        localization = jsonDirectory ?? JSONDirectory(name: "empty", files: [])
+    init(localizations: [String: Node]? = nil, defaultLocale: String? = nil) {
+        self.localizations = localizations ?? [:]
+        self.defaultLocale = defaultLocale ?? localizations?.keys.first ?? "default"
```
Comment (LoganWright) [inline Sources/Vapor/Localization/Localization.swift]: I'd rather just have `defaultLocale ?? "default"` here. The `keys.first` is essentially random and I think it'll make debugging more difficult.

File: Sources/Vapor/Localization/Localization.swift [hunk 5]
```diff
@@ -1,26 +1,52 @@
...
+            let data = try DataFile().load(path: path)
+            localizations[name] = try JSON(bytes: data).makeNode()
+        }
+        
+        self.init(localizations: localizations)
     }
 
-    init(jsonDirectory: JSONDirectory? = nil) {
-        localization = jsonDirectory ?? JSONDirectory(name: "empty", files: [])
+    init(localizations: [String: Node]? = nil, defaultLocale: String? = nil) {
```
Comment (LoganWright) [inline Sources/Vapor/Localization/Localization.swift]: public?

File: Sources/Vapor/View/LeafLocalizeTag.swift [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        // Validate the argument count
+        guard arguments.count > 0 else { throw Error.expectedAtLeasOneArgument }
+        
+        // Get the laguage
+        guard let language = arguments[0].value?.string else {
+            throw Error.invalidLanguage
+        }
+        
+        // Convert the path to an array of strings
+        let path = try arguments.dropFirst().map { e -> String in
```
Comment (LoganWright) [inline Sources/Vapor/View/LeafLocalizeTag.swift]: I'm not certain what `e` represents here, let's name it something more explicit

File: Sources/Vapor/View/LeafRenderer.swift [hunk 7]
```diff
@@ -2,7 +2,12 @@ import Leaf
 
 public final class LeafRenderer: ViewRenderer {
     public let stem: Stem
-
+    
+    public convenience init(viewsDir: String, localization: Localization) {
```
Comment (LoganWright) [inline Sources/Vapor/View/LeafRenderer.swift]: Localization needs to be optional here w/ default nil. `Localization? = nil`

File: Sources/Vapor/View/LeafRenderer.swift [hunk 8]
```diff
@@ -2,7 +2,12 @@ import Leaf
...
+        self.init(viewsDir: viewsDir)
+        stem.register(LocalizeTag(localization: localization))
```
Comment (LoganWright) [inline Sources/Vapor/View/LeafRenderer.swift]: w/ optional localization, could be:

```swift
localization.flatMap { stem.register(LocalizeTag(localization: $0)) }
```

File: Tests/VaporTests/ViewTests.swift [hunk 9]
```diff
@@ -22,6 +23,17 @@ class ViewTests: XCTestCase {
         let view = try View(bytes: "42".bytes)
         XCTAssertEqual(try view.makeBytes(), "42".bytes)
     }
+    
+    func testLeafLocalization() throws {
+        // Get the localization
+        let localization = LocalizationTests().localization
+        
+        // Create a localized Leaf renderer and check its value
+        let r = LeafRenderer(viewsDir: "ferret", localization: localization)
```
Comment (tannernelson) [inline Tests/VaporTests/ViewTests.swift]: lol yes ferret

File: Sources/Vapor/Localization/Localization.swift (hunk 2 above)
Comment (NathanFlurry) [inline Sources/Vapor/Localization/Localization.swift]: I thought so too originally, but having the `Localization` directory should be optional in all projects. If the user does not provide localizations, the program should still run just fine. Many web servers won't have localizations.

**Comment by tannernelson** [comment]
The localization changes are good, but I wonder how useful the Leaf tag will be. It doesn't solve the problem of dynamically changing the language the templates render in. You need some way to change from one language to another and have all templates render that language.

Have you looked at https://github.com/vapor/vapor/issues/546 at all?

This would allow the templates to change depending on what HTTP accept lang the request has.

File: Sources/Vapor/View/LeafRenderer.swift (hunk 8 above)
Comment (NathanFlurry) [inline Sources/Vapor/View/LeafRenderer.swift]: The more you know 🌈
I kind of wish it was named `exists` so you could do `localization.exists { ... }`. `flatMap` isn't very clear in my opinion, so I usually end up using `if let`.

**Comment by NathanFlurry** [comment]
@tannernelson Yeah, for now I would assume users would say something like `drop.view.make("welcome", [ "lang", request.acceptedLangauge ])`.

One solution would be to make middleware process requests twice; the first time with `ResponseRepresentable` data types and the second time with the actual `Response` data type. In addition, `ViewRenderer.make` would return a `ResponseRepresentable` object that holds the bytes for the raw, un-rendered view and the components. This way, you could write middleware that modifies the components for each view before rendering it. e.g. you could automatically add the "lang" component.

**Comment by LoganWright** [comment]
I do like the underlying localization api, I think we're all in agreement that we can polish that up and get it in.

The disagreement is how and where localization should fit in.

- Tanners suggestion has some advantages of being able to push more of the code to swift, possibly even creating variable extensions which increase type safety and reduce likelihood of typos.

- Nathan's suggestion is kind of a nice integration where you can declare your keys right into the template so there's less to do at the call sight.

Both have pros and cons, and I think both have a place. At the very least, if we decide LeafLocalization doesn't need to be in core behavior, we could whip up a provider pretty easily. (Side discussion would be a good way for people to share leaf tags).

**Comment by tannernelson** [comment]
Agreed. Could we move the Leaf localization part to a separate PR @NathanFlurry? That way we can get the other stuff merged ASAP.

**Comment by LoganWright** [comment]
@NathanFlurry do you have write access to bring this branch in? That way we can help keep it up to date.

**Comment by NathanFlurry** [comment]
@LoganWright I don't believe I have write access to do so.

**Comment by LoganWright** [comment]
@NathanFlurry weird, it looks like you should have permissions. I'll look into it, and no worries, I can 
bring it over tomorrow morn

**Comment by LoganWright** [comment]
Closed in preference of: #599

**Pull request closed without merging**